File: libraries/ndf/ndf_hsmod.f
      SUBROUTINE NDF_HSMOD( HMODE, INDF, STATUS )
*+
*  Name:
*     NDF_HSMOD

*  Purpose:
*     Set the history update mode for an NDF.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF_HSMOD( HMODE, INDF, STATUS )

*  Description:
*     The routine sets the mode to be used for updating the history
*     component of an NDF. This allows control over the amount of
*     history information subsequently recorded. It also allows history
*     recording to be disabled completely.

*  Arguments:
*     HMODE = CHARACTER * ( * ) (Given)
*        The history update mode required: 'DISABLED', 'QUIET',
*        'NORMAL' or 'VERBOSE'. This value may be abbreviated, to no
*        less than three characters. In addition, 'SKIP' may be supplied.
*        This is similar to 'DISABLED', in that no history record will
*        be added to the NDF when the NDF is closed. However, 'SKIP' makes
*        no permanent change to the update mode - the next time the NDF
*        is accessed it will have its original update mode.
*     INDF = INTEGER (Given)
*        NDF identifier.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Copyright:
*     Copyright (C) 1993 Science & Engineering Research Council
*     Copyright (C) 2008 Science & Technology Facilities Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     DSB: David S Berry (JAC, UCLan)
*     {enter_new_authors_here}

*  History:
*     18-MAY-1993 (RFWS):
*        Original version.
*     4-JUL-2008 (DSB):
*        Added SKIP mode.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_CONST'        ! NDF_ private constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Global Variables:
      INCLUDE 'NDF_DCB'          ! NDF_ Data Control Block
*        DCB_HLOC( NDF__MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator for NDF history component.
*        DCB_HUMOD( NDF__MXDCB ) = INTEGER (Write)
*           History recording update mode.

      INCLUDE 'NDF_ACB'          ! NDF_ Access Control Block
*        ACB_IDCB( NDF__MXACB ) = INTEGER (Read)
*           Index to data object entry in the DCB.

*  Arguments Given:
      CHARACTER * ( * ) HMODE
      INTEGER INDF

*  Status:
      INTEGER STATUS             ! Global status

*  External References:
      LOGICAL NDF1_SIMLR         ! String comparison with abbreviation

*  Local Variables:
      CHARACTER * ( NDF__SZHUM ) MODE ! Required mode
      CHARACTER * ( NDF__SZHUM ) TEXT ! UPDATE_MODE value
      INTEGER DIM( 1 )           ! Dummy dimension size array
      INTEGER HUM                ! History update mode code
      INTEGER IACB               ! Index of NDF entry in the ACB
      INTEGER IDCB               ! Index to data object entry in the DCB
      INTEGER NC                 ! Number of characters in TEXT
      LOGICAL SKIP               ! Was SKIP supplied?
*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the NDF identifier.
      CALL NDF1_IMPID( INDF, IACB, STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  If the history update mode is SKIP, set a flag and use DISABLED instead.
         SKIP = NDF1_SIMLR( HMODE, 'SKIP', 3 )
         IF( SKIP ) THEN
            MODE = 'DISABLED'
         ELSE
            MODE = HMODE
         END IF

*  Check that WRITE access to the NDF is available and validate the
*  history update mode string.
         CALL NDF1_CHACC( IACB, 'WRITE', STATUS )
         CALL NDF1_CHHUM( MODE, HUM, STATUS )

*  Obtain an index to the data object entry in the DCB and ensure that
*  DCB history information is available.
         IDCB = ACB_IDCB( IACB )
         CALL NDF1_DH( IDCB, STATUS )
         IF ( STATUS .EQ. SAI__OK ) THEN

*  Check that a history component is present and report an error if it
*  is not.
            IF ( DCB_HLOC( IDCB ) .EQ. DAT__NOLOC ) THEN
               STATUS = NDF__NOHIS
               CALL NDF1_DMSG( 'NDF', IDCB )
               CALL ERR_REP( 'NDF_HSMOD_NOHIS',
     :                       'There is no history component present ' //
     :                       'in the NDF structure ^NDF (possible ' //
     :                       'programming error).', STATUS )

*  Otherwise, test for each update mode value and assign the
*  appropriate UPDATE_MODE component string.
            ELSE IF ( HUM .EQ. NDF__HDISA ) THEN
               TEXT = 'DISABLED'
               NC = 8
            ELSE IF ( HUM .EQ. NDF__HQUIE ) THEN
               TEXT = 'QUIET'
               NC = 5
            ELSE IF ( HUM .EQ. NDF__HNORM ) THEN
               TEXT = 'NORMAL'
               NC = 6
            ELSE IF ( HUM .EQ. NDF__HVERB ) THEN
               TEXT = 'VERBOSE'
               NC = 7

*  If the update mode is not recognised, then report an error.
            ELSE
               STATUS = NDF__FATIN
               CALL MSG_SETI( 'HUM', HUM )
               CALL ERR_REP( 'NDF_HSMOD_HUM',
     :                       'Invalid history update mode code ' //
     :                       '(^HUM) encountered (internal ' //
     :                       'programming error).', STATUS )
            END IF

*  Ensure that an UPDATE_MODE component of the required type and shape
*  exists in the NDF history structure and write the new value to it.
*  Skip this bit if SKIP was supplied, so that no permanent change is made
*  to the NDF.
            IF ( STATUS .EQ. SAI__OK ) THEN
               IF( .NOT. SKIP ) THEN
                  CALL CMP_MODC( DCB_HLOC( IDCB ), 'UPDATE_MODE', NC,
     :                           0, DIM, STATUS )
                  CALL CMP_PUT0C( DCB_HLOC( IDCB ), 'UPDATE_MODE',
     :                            TEXT( : NC ), STATUS )
               END IF

*  Modify the corresponding DCB update mode value.
               IF ( STATUS .EQ. SAI__OK ) THEN
                  DCB_HUMOD( IDCB ) = HUM
               END IF
            END IF
         END IF
      END IF

*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'NDF_HSMOD_ERR',
     :   'NDF_HSMOD: Error setting the history update mode for an ' //
     :   'NDF.', STATUS )
         CALL NDF1_TRACE( 'NDF_HSMOD', STATUS )
      END IF

      END

File: libraries/ndf/ndfHsmod.c
#include <stdlib.h>
#include "sae_par.h"
#include "dat_par.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "ndf.h"
#include "star/util.h"
#include "star/cmp.h"
#include "mers.h"

void ndfHsmod_( const char *hmode, int indf, int *status ){
/*
*+
*  Name:
*     ndfHsmod

*  Purpose:
*     Set the history update mode for an NDF.

*  Synopsis:
*     void ndfHsmod( const char *hmode, int indf, int *status )

*  Description:
*     This function sets the mode to be used for updating the history
*     component of an NDF. This allows control over the amount of history
*     information subsequently recorded. It also allows history recording
*     to be disabled completely.

*  Parameters:
*     hmode
*        Pointer to a null terminated string holding the history update
*        mode required: "DISABLED", "QUIET", "NORMAL" or "VERBOSE". This
*        value may be abbreviated, to no less than three characters. In
*        addition, "SKIP" may be supplied. This is similar to "DISABLED",
*        in that no history record will be added to the NDF when the NDF is
*        closed. However, "SKIP" makes no permanent change to the update
*        mode - the next time the NDF is accessed it will have its original
*        update mode.
*     indf
*        NDF identifier.
*     *status
*        The global status.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     3-APR-2019 (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   NdfACB *acb;          /* Pointer to NDF entry in the ACB */
   NdfDCB *dcb;          /* Pointer to data object entry in the DCB */
   char mode[ NDF__SZHUM + 1 ];    /* Required mode */
   char text[ NDF__SZHUM + 1 ];    /* UPDATE_MODE value */
   hdsdim dim;           /* Dummy dimension size array */
   int hum;              /* History update mode code */
   int skip;             /* Was SKIP supplied? */
   size_t nc;            /* Number of characters in TEXT */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure the NDF library has been initialised. */
   NDF_INIT( status );

/* Import the NDF identifier. */
   ndf1Impid( indf, &acb, status );
   if( *status == SAI__OK ) {

/* If the history update mode is "skip", set a flag and use DISABLED instead. */
      skip = ndf1Simlr( hmode, 1, 0, "SKIP", 3 );
      if( skip ) {
         star_strlcpy( mode, "DISABLED", sizeof( mode ) );
      } else {
         star_strlcpy( mode, hmode, sizeof( mode ) );
      }

/* Check that WRITE access to the NDF is available and validate the
   history update mode string. */
      ndf1Chacc( acb, "WRITE", status );
      ndf1Chhum( mode, &hum, status );

/* Obtain an index to the data object entry in the DCB and ensure that
   DCB history information is available. */
      dcb = acb->dcb;
      ndf1Dh( dcb, status );
      if( *status == SAI__OK ) {

/* Check that a history component is present and report an error if it
   is not. */
         if( !dcb->hloc ) {
            *status = NDF__NOHIS;
            ndf1Dmsg( "NDF", dcb );
            errRep( " ", "There is no history component present in the NDF "
                    "structure ^NDF (possible programming error).", status );

/* Otherwise, test for each update mode value and assign the
   appropriate UPDATE_MODE component string. */
         } else if( hum == NDF__HDISA ) {
            star_strlcpy( text, "DISABLED", sizeof( text ) );
            nc = 8;
         } else if( hum == NDF__HQUIE ) {
            star_strlcpy( text, "QUIET", sizeof( text ) );
            nc = 5;
         } else if( hum == NDF__HNORM ) {
            star_strlcpy( text, "NORMAL", sizeof( text ) );
            nc = 6;
         } else if( hum == NDF__HVERB ) {
            star_strlcpy( text, "VERBOSE", sizeof( text ) );
            nc = 7;

/* If the update mode is not recognised, then report an error. */
         } else {
            *status = NDF__FATIN;
            msgSeti( "HUM", hum );
            errRep( " ", "Invalid history update mode code (^HUM) "
                    "encountered (internal programming error).", status );
         }

/* Ensure that an UPDATE_MODE component of the required type and shape
   exists in the NDF history structure and write the new value to it.
   Skip this bit if "skip" was supplied, so that no permanent change is made
   to the NDF. */
         if( *status == SAI__OK ) {
            if( !skip ) {
               cmpModC( dcb->hloc, "UPDATE_MODE", nc, 0, &dim, status );
               cmpPut0C( dcb->hloc, "UPDATE_MODE", text, status );
            }

/* Modify the corresponding DCB update mode value. */
            if( *status == SAI__OK ) dcb->humod = hum;
         }
      }
   }

/* If an error occurred, then report context information and call the
   error tracing function. */
   if( *status != SAI__OK ) {
      errRep( " ", "ndfHsmod: Error setting the history update mode for an "
              "NDF.", status );
      ndf1Trace( "ndfHsmod", status );
   }

/* Restablish the original AST status pointer */
   NDF_FINAL

}

